### PR TITLE
Fix handling of internal libraries.

### LINKED ISSUE #51: Confusing handling of Cabal internal libraries
[Our project](https://github.com/amnh/pcg/) uses [multiple packages as sub-libraries](https://github.com/amnh/PCG/tree/master/lib), one of which has [multiple internal libraries](https://github.com/amnh/PCG/tree/master/lib/pcg-core) that are re-exported.

When we run `weeder` on the package with multiple internal libraries, we get the following output:

```
= Package pcg-core =

== Section library ==
Module not compiled
* Analysis.Parsimony.Additive
* Analysis.Parsimony.Additive.Internal
* Analysis.Parsimony.Dynamic.DirectOptimization
* Analysis.Parsimony.Dynamic.DirectOptimization.Internal
* Analysis.Parsimony.Dynamic.DirectOptimization.Pairwise
* Analysis.Parsimony.Dynamic.DirectOptimization.Pairwise.FFI
* Analysis.Parsimony.Dynamic.DirectOptimization.Pairwise.Internal
* Analysis.Parsimony.Dynamic.DirectOptimization.Pairwise.NeedlemanWunsch
* Analysis.Parsimony.Dynamic.DirectOptimization.Pairwise.Ukkonen
* Analysis.Parsimony.Dynamic.DirectOptimization.Pairwise.Ukkonen.Internal
* Analysis.Parsimony.Dynamic.DirectOptimization.Pairwise.Ukkonen.Ribbon
* Analysis.Parsimony.Fitch
* Analysis.Parsimony.Fitch.Internal
* Analysis.Parsimony.Sankoff
* Analysis.Parsimony.Sankoff.Internal
* Analysis.Scoring
* Analysis.TotalEdgeCost
* Bio.Character
* Bio.Character.Decoration.Additive
* Bio.Character.Decoration.Additive.Class
* Bio.Character.Decoration.Additive.Internal
* Bio.Character.Decoration.Continuous
* Bio.Character.Decoration.Continuous.Class
* Bio.Character.Decoration.Continuous.Internal
* Bio.Character.Decoration.Discrete
* Bio.Character.Decoration.Dynamic
* Bio.Character.Decoration.Dynamic.Class
* Bio.Character.Decoration.Dynamic.Internal
* Bio.Character.Decoration.Fitch
* Bio.Character.Decoration.Fitch.Class
* Bio.Character.Decoration.Fitch.Internal
* Bio.Character.Decoration.Metric
* Bio.Character.Decoration.Metric.Class
* Bio.Character.Decoration.Metric.Internal
* Bio.Character.Decoration.NonMetric
* Bio.Character.Decoration.NonMetric.Class
* Bio.Character.Decoration.NonMetric.Internal
* Bio.Character.Decoration.Shared
* Bio.Character.Encodable
* Bio.Character.Encodable.Continuous
* Bio.Character.Encodable.Continuous.Class
* Bio.Character.Encodable.Continuous.Internal
* Bio.Character.Encodable.Dynamic
* Bio.Character.Encodable.Dynamic.Class
* Bio.Character.Encodable.Dynamic.Internal
* Bio.Character.Encodable.Internal
* Bio.Character.Encodable.Static
* Bio.Character.Encodable.Static.Class
* Bio.Character.Encodable.Static.Internal
* Bio.Character.Encodable.Stream
* Bio.Character.Type
* Bio.Graph
* Bio.Graph.BinaryRenderingTree
* Bio.Graph.Component
* Bio.Graph.Constructions
* Bio.Graph.Forest
* Bio.Graph.LeafSet
* Bio.Graph.Node
* Bio.Graph.Node.Context
* Bio.Graph.Node.Internal
* Bio.Graph.PhylogeneticDAG
* Bio.Graph.PhylogeneticDAG.Class
* Bio.Graph.PhylogeneticDAG.DynamicCharacterRerooting
* Bio.Graph.PhylogeneticDAG.Internal
* Bio.Graph.PhylogeneticDAG.NetworkEdgeQuantification
* Bio.Graph.PhylogeneticDAG.Postorder
* Bio.Graph.PhylogeneticDAG.Preorder
* Bio.Graph.PhylogeneticDAG.Reification
* Bio.Graph.ReferenceDAG
* Bio.Graph.ReferenceDAG.Internal
* Bio.Graph.ReferenceDAG.Network
* Bio.Graph.ReferenceDAG.Traversal
* Bio.Graph.ReferenceDAG.Utility
* Bio.Graph.Solution
* Bio.Graph.ZipperDAG
* Bio.Graph.ZipperDAG.Internal
* Bio.Metadata
* Bio.Metadata.CharacterName
* Bio.Metadata.Continuous
* Bio.Metadata.Discrete
* Bio.Metadata.Discrete.Class
* Bio.Metadata.Discrete.Internal
* Bio.Metadata.DiscreteWithTCM
* Bio.Metadata.DiscreteWithTCM.Class
* Bio.Metadata.DiscreteWithTCM.Internal
* Bio.Metadata.Dynamic
* Bio.Metadata.Dynamic.Class
* Bio.Metadata.Dynamic.Internal
* Bio.Metadata.General
* Bio.Metadata.General.Class
* Bio.Metadata.General.Internal
* Bio.Sequence
* Bio.Sequence.Block
* Bio.Sequence.Block.Builder
* Bio.Sequence.Block.Character
* Bio.Sequence.Block.Internal
* Bio.Sequence.Block.Metadata
* Bio.Sequence.Character
* Bio.Sequence.Internal
* Bio.Sequence.Metadata
* Data.EdgeLength
* Data.EdgeSet
* Data.MetricRepresentation
* Data.NodeLabel
* Data.TCM
* Data.TCM.Dense
* Data.TCM.Dense.FFI
* Data.TCM.Internal
* Data.TopologyRepresentation
* Test.Custom.NucleotideSequence
Redundant build-depends entry
* 
* QuickCheck
* bimap
* binary
* bv-little
* cassava
* compact
* containers
* data-default
* deepseq
* dlist
* foldl
* graphviz
* hashable
* keys
* lens
* monad-loops
* mtl
* parallel
* pcg-alphabet
* pcg-analysis
* pcg-data-structures
* pcg-evaluation
* pcg-exportable
* pcg-serialize
* pcg-tcm
* pcg-tcm-memo
* pcg-utility
* pretty-tree
* semigroupoids
* smallcheck
* text
* text-short
* text-show
* text-show-instances
* unordered-containers
* vector
* vector-binary-instances
* vector-instances
* xml

== Section test:analysis-tests ==
Redundant build-depends entry
* 
* pcg-analysis
* pcg-data-structures
* pcg-tcm

== Section test:data-structures-tests ==
Redundant build-depends entry
* 
* pcg-data-structures

== Section test:tcm-tests ==
Redundant build-depends entry
* 
* pcg-tcm
* pcg-tcm-memo

= Package pcg-language =

== Section library ==
Redundant build-depends entry
* 
```

`weeder` reports as weeds that the [re-exported](https://www.haskell.org/cabal/users-guide/developing-packages.html#pkg-field-library-reexported-modules) modules were not compiled, however these modules need to be re-exported from the package's main library as the interface to be consumed by other packages in our code base.

Additionally, `weeder` reports the [internal libraries](https://www.haskell.org/cabal/users-guide/developing-packages.html#sublibs) which are dependencies of other build targets as redundant, however they are require to build the build target and cannot be removed.

I'm hoping to get your input on whether or not these weed reports are expected, they seem to be erroneous from my perspective in this use case.

### PR DESCRIPTION
**Pull request by kindaro**
Resolve #51.

Special thanks to @rvl for test case #49, which is included here.

### CODE REVIEW

**Comment by kindaro** [comment]
@ndmitchell  I understand that you are not working on this project anymore. However, it is still useful. [The other version][ocharles] requires GHC to understand `-fwrite-ide-info`, which is only available from 8.8 upwards. There is any number of projects that only build with, say, GHC 8.6. It so happens that I am working on one such project, and I am enjoying the benefits of Weeder with this patch presently. So maybe you would still like to consider merging for the benefit of others who find themselves in a situation like mine.

[ocharles]: https://github.com/ocharles/weeder

File: test/.weeder.yaml [hunk 1]
```diff
@@ -77,3 +77,24 @@
       - module:
         - name: Dir.Unused
         - identifier: unused
+
+## TODO: These false positives are due to the use of Cabal internal
+## libraries, which are not yet supported by Weeder.
```
Comment (ndmitchell) [inline test/.weeder.yaml:82]: Is this comment true after this patch is merged?

**Review by ndmitchell — COMMENTED** [review]
Sure, I'll merge. Conflicted on whether I make a release, as by now the CI infrastructure has rotted, so my confidence in any release will be correspondingly lower. But if you promise you've run the tests I guess I will?

File: test/.weeder.yaml (hunk 1 above)
Comment (kindaro) [reply test/.weeder.yaml:82]: Good question, I have no idea!

**Comment by kindaro** [comment]
I must confess, I have not been able to figure out how to automatically run the test suite. There is no _«test»_ clause in `weeder.cabal`, and the directory `test` does not contain any test execution engine either. So my verification went only as far as running the pristine and the patched version on the examples in `test`, as well as on some of my private projects, and comparing the output.

So, if you can give me a hint as to how the test suite is supposed to work?

In the meanwhile, I am going to invest an evening into making the continuous integration green, at least Travis. It looks like it only needs an adjustment to the `tested-with` Cabal clause.

**Comment by ndmitchell** [comment]
Thanks!

**Comment by ndmitchell** [comment]
I released 1.0.9.

I actually often put the executable as its own checker for tools. Makes it way easier to integrate more closely - I do the same for hlint.

**Pull request merged**